### ISSUE TITLE
Compute_at schedule generates a kernel that doesn't compute anything (arm-64-osx-metal)

### ISSUE THREAD

**Issue by yyuting** [halide/Halide#7770]
Attaching below the generator that triggers this behavior with `extra_compute=true` and `extra_schedule=true`.

The generator simply copies input to output. But when `extra_compute=true`, it adds `masks_up_y` (which is always 0) to the output.

With the additional null computation and an extra `compute_at` schedule, the kernel becomes 10x faster, but nothing is written to the output buffer (output buffer still has its initialized value). 

The generated kernel behaves correctly without the `compute_at` schedule.

I'm using Mac M2 with `target=host-metal`.

This behavior happens several times when I want to use `compute_at` for the intermediate stages of upsampling operators. And this only happens on the metal GPU targets, CPU targets are fine. Do you have any suggestions for a workaround? Thanks a lot!

```
#include "Halide.h"

namespace {

using namespace Halide;  // NOLINT(build/namespaces)

Var x("x");
Var y("y");
Var z("z");
Var c("c");

class Debug : public Generator<Debug> {
 public:
  GeneratorParam<bool> extra_compute{"extra_compute", false};
  GeneratorParam<bool> extra_schedule{"extra_schedule", false};

  // Input image.
  Input<Buffer<float>> input{"input", 3};

  Output<Buffer<float>> output{"output", 3};

  void generate() {
    // These functions have no effect to the final output value, but will
    // trigger the scheduling bug
    Func masks_up_x("masks_up_x");
    Func masks_up_y("masks_up_y");

    masks_up_x(x, y, z) = 0.f;

    Expr yv = cast<float>(y) - 0.5f;
    Expr yi = cast<int>(floor(yv));
    masks_up_y(x, y, z) = masks_up_x(x, yi, z);

    if (extra_compute) {
      // Add masks_up_y, which should not change the output at all
      output(x, y, c) = input(x, y, c) + masks_up_y(x, y, 0);
    } else {
      output(x, y, c) = input(x, y, c);
    }

    // Schedule
    if (!using_autoscheduler()) {
      if (get_target().has_gpu_feature()) {
        Var xi, xo, yi, yo;
        output.compute_root()
            .reorder(c, x, y)
            .bound(c, 0, 3)
            .unroll(c)
            .gpu_tile(x, y, xo, yo, xi, yi, 8, 32);

        if (extra_schedule) {
          // This schedule generates a Metal GPU kernel that does not compute anything
          masks_up_x.compute_at(output, xo).unroll(z).gpu_threads(x, y);
        }

      } else {
      }
    }

    // Estimates
    {}
  }
};
}  // namespace

HALIDE_REGISTER_GENERATOR(Debug, debug)

```

**Comment by abadams** [comment]
I'm not very familiar with metal. It appears there's a debug-only validation of the thread-group size:

https://github.com/halide/Halide/blob/main/src/runtime/metal.cpp#L775

and then dispatch_threadgroups itself doesn't have any way to signal an error. Maybe the idea is that the error is surfaced on the next context acquire? @shoaibkamil do you know what this should look like?

**Comment by abadams** [comment]
I think you want .bound_extent(x, 8).bound_extent(y, 32)

I'm surprised the system let you use the free Var xo in a bound expression - that's not supposed to be legal

**Comment by yyuting** [comment]
Thanks, Andrew. 
I can use `.bound_extent(x, 8)` fine, but still get the `halide_error_explicit_bounds_too_small` error for `.bound_extent(y, 32)`. 

To make things simpler, I also explored changing `yi` to be equivalent to `y`:
```
    Expr yv = cast<float>(y);
    Expr yi = cast<int>(floor(yv));
    masks_up_y(x, y, z) = masks_up_x(x, yi, z);
```
But `.bound_extend(y, 32)` still fails. Is Halide still doing some safety checks that I didn't specify explicitly?

**Comment by abadams** [comment]
For bound_extent(y, 32) I get an internal compiler error:

Internal Error at /Users/abadams/projects/Halide/src/CodeGen_Metal_Dev.cpp:213 triggered by user code at :
Condition failed: !function_takes_user_context(op->name):
halide_error_explicit_bounds_too_small

Looks like the metal backend can't handle any runtime functions, and rather than return a sensible compile-time error message, it just panics, so bound_extent can't work (because it requires an assert in cases where it can't statically prove things). Hrm. It compiles and runs fine with -no_asserts, but that's not very satisfying.

The root thing that's making all this messy is that Halide's bounds inference doesn't reason well about floating point math. If it's possible to do your resampling with integer math (e.g. if the upsample is by a compile-time-constant integer factor), do that instead. 

Alternatively, don't compute mask_up_x per gpu block. The complex relationship between its bounds and the bounds of the output make it a problematic schedule. If the upsampling factor is large I'd just run mask_up_x as its own gpu_tile'd compute_root stage.

**Comment by yyuting** [comment]
Thanks, Andrew! That's very helpful!

**Comment by abadams** [comment]
Closed in favor of #7780
